zilencer/lib/remote_counts.py

```diff
@@ -0,0 +1,85 @@
+import datetime
+
+from django.db import connection
+from django.utils.timezone import now as timezone_now
+from psycopg2.sql import SQL, Literal
+
+from zilencer.models import RemoteInstallationCount, RemoteZulipServer
+
+
+class MissingDataError(Exception):
+    pass
+
+
+def compute_max_monthly_messages(remote_server: RemoteZulipServer) -> int:
+    # Calculate the maximum amount of messages that the server had within a month.
+    # out of the last 3 months.
+
+    # Filtering on property is a placeholder here; it
+    # currently doesn't change the result. Note that we can't
+    # filter on the actual property we care about
+    # ('messages_sent:message_type:day'), because zeros are not
+    # recorded.
+    # For non-legacy servers, we probably want to distinguish not uploading
+    # data vs. uploading minimal billing data (and thus not qualified for the
+    # Starter plan) vs. uploading all the data.
+    if not RemoteInstallationCount.objects.filter(
+        server=remote_server,
+        property="active_users_audit:is_bot:day",
+        end_time__lte=timezone_now() - datetime.timedelta(days=3),
+    ).exists():
+        raise MissingDataError
+
+    query = SQL(
+        """
+    WITH server_message_stats_daily AS -- Up to 4 rows per day for different subgroups
+    (
+        SELECT
+            r.end_time,
+            r.value AS message_cnt
+        FROM
+            zilencer_remoteinstallationcount r
+        WHERE
+            r.property = 'messages_sent:message_type:day'
+            AND end_time >= CURRENT_TIMESTAMP(0) - INTERVAL '90 days'
+            AND r.server_id = {server_id}
+    ),
+    server_message_stats_monthly AS (
+        SELECT
+            CASE
+                -- date_trunc('month', current_timestamp(0)) gives us the start
+                -- of the current month. We may want to replace that with the start
+                -- of the billing cycle when we have that.
+                WHEN end_time > date_trunc('month', current_timestamp(0)) THEN 0
+                WHEN date_trunc('month', current_timestamp(0)) - end_time <= INTERVAL '30 days' THEN 1
+                WHEN date_trunc('month', current_timestamp(0)) - end_time <= INTERVAL '60 days' THEN 2
+            END AS billing_month,
```

Comment:
@alya Is this the type of thing we wanna do here? Or just count by the more obvious `(today-30days, today)`. `(today-60days, today-30days)`, `(today-90days, today-60days)` intervals?

Reply:
Do we consider billing cycles to start at the moment of the upgrade, or do we round to the nearest day?

I'm not sure, but having the times be precise might be easier to understand for any orgs with really spiky traffic (like events) that are looking at their stats?

Reply:
It looks like we truncate, so that eventually nobody has a billing cycle date later than the 28th of the month.

It may be easier to use 30-day periods for the purpose of this calculation, rather than monthly billing cycles, just because that's a lot easier to explain concisely?

```
# Be extremely careful changing this function. Historical billing periods                               
# are not stored anywhere, and are just computed on the fly using this                                  
# function. Any change you make here should return the same value (or be                                
# within a few seconds) for basically any value from when the billing system                            
# went online to within a year from now.                                                                
def add_months(dt: datetime, months: int) -> datetime:                                                  
    assert months >= 0                                                                                  
    # It's fine that the max day in Feb is 28 for leap years.                                           
    MAX_DAY_FOR_MONTH = {                                                                               
        1: 31,                                                                                          
        2: 28,                                                                                          
        3: 31,                                                                                          
        4: 30,                                                                                          
        5: 31,                                                                                          
        6: 30,                                                                                          
        7: 31,                                                                                          
        8: 31,                                                                                          
        9: 30,                                                                                          
        10: 31,                                                                                         
        11: 30,                                                                                         
        12: 31,                                                                                         
    }                                                                                                   
    year = dt.year                                                                                      
    month = dt.month + months                                                                           
    while month > 12:                                                                                   
        year += 1                                                                                       
        month -= 12                                                                                     
    day = min(dt.day, MAX_DAY_FOR_MONTH[month])                                                         
    # datetimes don't support leap seconds, so don't need to worry about those                          
    return dt.replace(year=year, month=month, day=day)                                                  
```

Reply:
Ok, yeah, if we effectively store a billing day rather than a billing time, then we should round the logic in this query to days too.

Reply:
Changed this to a simple
```
            CASE
                WHEN current_timestamp(0) - end_time <= INTERVAL '30 days' THEN 0
                WHEN current_timestamp(0) - end_time <= INTERVAL '60 days' THEN 1
                WHEN current_timestamp(0) - end_time <= INTERVAL '90 days' THEN 2
```

Reply:
I'm now wondering if rounding to the nearest day before subtracting would actually be better than not. Will let @timabbott comment.